<table align="left" width=100%>
    <tr>
        <td width="20%">
            <img src="faculty.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                  <b> Faculty Notebook <br> Session 2) </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## Table of Content

1. **[ Examples of Conditional Statement and Loops](#examples)**
    - 1.1 - [if statement](#if_statement)
    - 1.2 - [if else statement](#if_else)
    - 1.3 - [nested conditions](#if_elif_else)
    - 1.4 - [for loop and nested for loop](#for_and_nested)
    - 1.5 - [while loop](#while_loop)
2. **[ What are List Comprehensions](#List_Comprehensions)**
    - 2.1 - [for loop](#for_loop_list)
    - 2.2 - [nested for loop](#nested_for_loop_list)
    - 2.3 - [if statement](#if_list)
    - 2.4 - [nested condition](#nested_condition) 

3. **[Built-in Functions](#builtin)**

4. **[User Defined Functions](#user)**
    - 4.1 - **[User defined functions without arguments](#without_arg)**
    - 4.2 - **[User defined functions with arguments](#with_arg)**
    - 4.3 - **[Recursive Functions](#Recursive_Functions)**

5. **[Scope of  Variables](#scope)**

6. **[Lambda Functions](#lambda)**

7. **[Overview of map, reduce and filter functions](#overview)**
    - 7.1 - [Function map()](#map)
    - 7.2 - [Function filter()](#filter)
    - 7.3 - [Function reduce()](#reduce)
    - 7.4 - [Function accumulate() ](#accumulate)




<a id="examples"> </a>
## 1. Examples of Conditional Statement and Loops

<a id="if_statement"> </a>
## 1.1 if statement

Now, we will see some examples on if statements

**Check whether the number is smaller than 4738 or not**

In [1]:
Val = 3636
if(Val<4738):
    print("number is smaller than 4738")

number is smaller than 4738


In the above example, we declared a variable called ‘Val’ with the value as 3636 and in the if statement we are checking if the number is less than 4738 or not. If the condition is true then a set of statements inside the 'if' block will be executed

**Check whether the number is positive or not**

In [2]:
Val = 453
if Val > 0:
    print(Val, "is a positive number")

453 is a positive number


<a id="if_else"> </a>
## 1.2 if else statement

Let’s see an example of the implementation of the if…else statement

**Check whether the given number is positive or not**

In [3]:
a = -5

if a >= 0:
    print("The number is positive or zero")
else:
    print("The number is negative")

The number is negative


**Check whether the number is even or odd**

In [4]:
num = 22
if num % 2 == 0:
    print("The number is even")
else:
    print("The number is odd")

The number is even


<a id="if_elif_else"> </a>
## 1.3 nested conditions

**Let’s take an example to check whether a number is odd and greater than 20**

In [5]:
x = 45
if x%2 != 0:
    if x > 20:
        print("This number is odd and is greater than 20")
    else:
        print("This is number is odd but not greater than 20")
else:
    print("The number is not even")

This number is odd and is greater than 20


**Create a simple calculator using if elif condition**

In [6]:
Value1 = float(input("Enter first number: "))
op = input("Enter operator: ")
Value2 = float(input("Enter second number: "))
if op == '+':
    print(Value1 + Value2)
elif op == '-':
    print(Value1 - Value2)
elif op == '/':
    print(Value1 / Value2)
elif op == '*':
    print(Value1 * Value2)
else:
    print("Invalid Operator")

Enter first number: 7
Enter operator: +
Enter second number: 2
9.0


**Example of nested if statement**

In [7]:
person_age = int(input(" Enter your age here:  "))
if person_age < 18:
    print(" You are not eligible to work ") 
else:
    if person_age >= 18 and person_age <= 60:
        print(" You are eligible to work ")
        print(" Please fill your details and apply")
    else:
        print(" You are too old to work as per the government rules")
        print(" Please collect your pension!")

 Enter your age here:  23
 You are eligible to work 
 Please fill your details and apply


In this program, you can enter your age and store it in the variable 'person_age'. If the age is less than 18, print the result for that condition. If the condition is not true, we check one more condition (Nested) and if it is true, we print result for that condition. If the nested condition fails, we print the other statement.

**Program to display the grade of the student**

In [1]:
Marks = int(input(" Enter Your Marks Here:  "))

if Marks >= 90:
    print("A grade")

elif Marks >=80:
    print("B grade")

elif Marks >=70:
    print("C grade")

elif Marks >= 65:
    print("D grade")

else:
    print("F grade")

D grade


Now, there are four possible outputs that can occur once we run the program:<br>
1. If the marks is greater than 90, the program will print A grade, if the marks is less than 90, the program will continue to the next statement<br>
2. If the marks is greater than or equal to 80, the program will print B grade, if the marks is 79 or less, the program will continue to the next statement<br>
3. If the marks is greater than or equal to 70, the program will print C grade, if the marks is 69 or less, the program will continue to the next statement<br>
4. If the marks is greater than or equal to 65, the program will print D grade, if the marks is 64 or less, the program will continue to the next statement<br>
5. The program will print F grade because all of the above conditions were not met

**Enter the amount and calculate discount based on the amount and given discount rate**

In [5]:
amt = int(input("Enter Sale Amount: "))
# checking conditions and calculating discount
if(amt>0):
    if amt<=5000:
        disc = amt*0.05
    elif amt<=15000:
        disc=amt*0.12
    elif amt<=25000:
        disc=0.2 * amt
    else:
         disc=0.3 * amt
    print("Discount : ",disc)
    print("Net Pay  : ",amt-disc)
else:
    print("Invalid Amount")


Discount :  3000000.0
Net Pay  :  7000000.0


**Find largest number among three numbers**

In [10]:
num1 = float(input("Enter first number: "))
num2 = float(input("Enter second number: "))
num3 = float(input("Enter third number: "))
 
if (num1 > num2) and (num1 > num3):
    largest = num1
    print("The largest number is",largest)
elif (num2 > num1) and (num2 > num3):
    largest = num2
    print("The largest number is",largest)
else:
    largest = num3
    print("The largest number is",largest)



Enter first number: 6
Enter second number: 32
Enter third number: 4
The largest number is 32.0


**Find the sum of natural numbers up to n**

In [6]:
# sum of natural numbers
num = 16

if num < 0:
    print("Enter a positive number")
else:
    sum = 0
    while(num > 0):
        sum += num
        num -= 1
    print("The sum is",sum)

The sum is 136


<a id="for_and_nested"> </a>
## 1.4 for loop and nested for loop

**Calculate the square of each element in the list**

In [12]:
list_1 = [1,3,13,91,100,200]
for i in list_1:
    print(i*i)

1
9
169
8281
10000
40000


**Write a code to print table of given number**

In [13]:
n=int(input())
for i in range(1,11):
    print(n,"x",i,"=",n*i)

6
6 x 1 = 6
6 x 2 = 12
6 x 3 = 18
6 x 4 = 24
6 x 5 = 30
6 x 6 = 36
6 x 7 = 42
6 x 8 = 48
6 x 9 = 54
6 x 10 = 60


**Let’ see now how to print full pyramid and asterisk(star) patterns in python**

In [14]:
print("Print full triangle using stars")
size = 5
m = (2 * size) - 2
for i in range(0, size):
    for j in range(0, m):
        print(end=" ")
    m = m - 1 # decrementing m after each loop
    for j in range(0, i + 1):
        # printing full Triangle pyramid using stars
        print("* ", end=' ')
    print(" ")

Print full triangle using stars
        *   
       *  *   
      *  *  *   
     *  *  *  *   
    *  *  *  *  *   


**Form a lower tringular matrix in ' * ' pattern**

In [15]:
for i in range(0, 5):
    for j in range(0, i+1):
        print("* ",end="")
    print()

* 
* * 
* * * 
* * * * 
* * * * * 


**Write a program to print alphabetical pattern using ascii values**

In [16]:
for i in range (65,70):
    # inner loop
    for j in range(65,i+1):
        print(chr(j),end="")
    print()

A
AB
ABC
ABCD
ABCDE


**Access each individual item of the sublists using nested for loop**

In [7]:
list_of_lists = [['Pune', 'Mumbai', 'Delhi'],[23, 12, 32],[9.29, 8.28, 7.27]]

for list in list_of_lists:
    for item in list:
        print(item)

Pune
Mumbai
Delhi
23
12
32
9.29
8.28
7.27


**Read a number n and print an identity matrix of the same order**

In [9]:
# identity matrix : A matrix with nxn and all Diagonal elements as 1
n=int(input("Enter a number: "))
for i in range(0,n):
    for j in range(0,n):
        if(i==j):
            print("1",sep=" ",end=" ")
        else:
            print("0",sep=" ",end=" ")
    print()

1 0 0 0 0 
0 1 0 0 0 
0 0 1 0 0 
0 0 0 1 0 
0 0 0 0 1 


**Access each individual letter from the given list of words**

In [19]:
words= ["Apple", "Banana", "Car", "Dolphin" ]
for word in words:
        #access word from the list
        print ("The following lines will print each letters of "+word)
        for letter in word:
                #access letter for the word
                print (letter)
        print("")

The following lines will print each letters of Apple
A
p
p
l
e

The following lines will print each letters of Banana
B
a
n
a
n
a

The following lines will print each letters of Car
C
a
r

The following lines will print each letters of Dolphin
D
o
l
p
h
i
n



**Print a number pattern using a for loop and range function**

In [20]:
# 15 is the total number to print
for num in range(15):
    for i in range(num):
        print (num, end=" ") #print number
    # new line after each row to display pattern correctly
    print("\n")



1 

2 2 

3 3 3 

4 4 4 4 

5 5 5 5 5 

6 6 6 6 6 6 

7 7 7 7 7 7 7 

8 8 8 8 8 8 8 8 

9 9 9 9 9 9 9 9 9 

10 10 10 10 10 10 10 10 10 10 

11 11 11 11 11 11 11 11 11 11 11 

12 12 12 12 12 12 12 12 12 12 12 12 

13 13 13 13 13 13 13 13 13 13 13 13 13 

14 14 14 14 14 14 14 14 14 14 14 14 14 14 



**Find the first 3-digit number that’s divisible by 12**

In [14]:
for i in range(0, 1000, 15):  # 3rd value is step value
    if len(str(i)) == 3:
        print(i)
        break

105


**Write a programm to display even numbers using continue statement**

In [15]:
for num in [230, 211, 932, 626, 234, 829, 424]:
    # Skipping the iteration when number is even
    if num%2 == 0:
        continue
    # This statement will be skipped for all even numbers
    print(num)

211
829


<a id="while_loop"> </a>
## 1.5 while loop

In [23]:
# while loop with if else
a = 7
while a > 0:
    a = a - 1
    if a == 2:
        break
    print(a)
else:
    print("Loop is finished")

6
5
4
3


In [24]:
# break in while
i = 1
while i < 6:
  print(i)
  if i == 3:
    break
  i += 1

1
2
3


In [16]:
# break in while loop
while True:
    n = input("Please enter 'hello':")
    if n.strip() == 'hello':
        break

In [17]:
# break in while loop
num_sum = 0
count = 0
while(count<10):
    num_sum = num_sum + count
    count = count + 1 
    if count== 4:
        break
print("Sum of first ",count,"integers is: ", num_sum)

Sum of first  4 integers is:  6


In [27]:
# continue in while loop
i = 0
while i < 7:
  i += 1
  if i == 4:
    continue
  print(i)

1
2
3
5
6
7


In [28]:
# continue in while loop
for x in range(7):
    if (x == 3 or x==6):
        continue
    print(x)

0
1
2
4
5


<a id="List_Comprehensions"> </a>
## 2. What are List Comprehensions?

<table align="left">
    <tr>
        <td width="15%">
            <img src="note.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>List comprehensions provide us with a simple way to create a list based on some iterable. During the creation, elements from the iterable can be conditionally included in the new list and transformed as needed.<br>
                        <br>Components of list comprehensions:<br>
                        . Output expression<br>
                        . Iterable<br>
                        . Iterator variable which represents the members of the iterable
                        
                        <img src="list.png", height="400", width="400">
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

**In python, list comprehensions are constructed as follows:**<br>
    
    list = [x for x in iterable]

<a id="List_Comprehensions_loops"> </a>
## List Comprehensions vs loops

<a id="for_loop_list"> </a>
### 2.1. for loop

Let’s look at an example that creates a list based on a string:<br>
Suppose, you want to separate the letters of the word 'Great Lakes' and add the letters as items of a list. The first thing that comes in mind would be using for loop<br>

In [29]:
# Using loop for constructing output list 
list_1=[]
for i in 'Great Lakes':
    list_1.append(i)
list_1

['G', 'r', 'e', 'a', 't', ' ', 'L', 'a', 'k', 'e', 's']

Here, square brackets signifies that the output is a list. list_1.append(i) is the function executed for each element and for i in 'Great Lakes' is used to iterate over each element. But with a python list comprehension, we can do this in one line

<table align="left">
    <tr>
        <td width="7%">
            <img src="caution.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>Remember that a python string is iterable. We can not iterate dictionary like this. Iterating through a dictionary gives us 'dict' object has no attribute 'append'
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [30]:
# Using loop for constructing output dictionary
dict_1={}
for i in 'Great Lakes':
    dict_1.append(i)
dict_1

AttributeError: 'dict' object has no attribute 'append'

In [31]:
# using list comprehensions
new_list = [i for i in 'Great Lakes']
print(new_list)

['G', 'r', 'e', 'a', 't', ' ', 'L', 'a', 'k', 'e', 's']


In the above example, a new list is assigned to variable 'new_list', and list contains the items of the iterable string 'Great Lakes'. We call print() function to print the output

**Find the squares of each element using list comprehension**

In [32]:
num = [13, 32, 33, 24]
num_squares = [n**2 for n in num]

print(num_squares)

[169, 1024, 1089, 576]


In the above example, square brackets signifies that the output is a list. n**2 is the expression executed for each element and for n in numbers is used to iterate over each element

**We can use list comprehension to iterate over list of strings. Convert all the words in the given list to the upper case**    

In [20]:
list_1 = ["yes", "it's", "a", "Beautiful", "Day"]

upper_list_1 = [str.upper() for str in list_1]

print(upper_list_1)

['YES', "IT'S", 'A', 'BEAUTIFUL', 'DAY']


**List comprehensions can be used to produce a list of a list, as shown below:**

In [21]:
list_1 = [1, 2, 3, 4]

square_cube_num = [ [a**2, a**3] for a in list_1]

print(square_cube_num)

[[1, 1], [4, 8], [9, 27], [16, 64]]


Here we use list_1 to create square_cube_num, which is a list of list containing squares and cubes of the numbers from list_1

<a id="nested_for_loop_list"> </a>
### 2.2 nested for loop

** Print the tables of numbers from 13 to 17**

In [22]:
[[i*j for j in range(1,11)] for i in range(13,18)]

[[13, 26, 39, 52, 65, 78, 91, 104, 117, 130],
 [14, 28, 42, 56, 70, 84, 98, 112, 126, 140],
 [15, 30, 45, 60, 75, 90, 105, 120, 135, 150],
 [16, 32, 48, 64, 80, 96, 112, 128, 144, 160],
 [17, 34, 51, 68, 85, 102, 119, 136, 153, 170]]

**Create a list of tuples with a and b such that a is less than b**

In [23]:
list_1 = [13, 72, 27, 83, 89, 54]
list_2 = [25, 37, 35]

new_list = [(a, b) for a in list_1 for b in list_2 if a < b]

print(new_list)

[(13, 25), (13, 37), (13, 35), (27, 37), (27, 35)]


### Conditional list comprehension
List comprehensions can utilize conditional statements to modify existing lists or other sequential data types when creating new lists.<br>
Conditional list comprehension adds only those items to the list that meet the condition (for which the condition is True).

**if statement**                        <img src="condition.png", height="550", width="550">

**if-else statement**                        <img src="ifelse.png", height="550", width="550">

<a id="if_list"> </a>
### 2.3 if satement

We create an example that uses mathematical operators, integers, and the range() sequence type. Take the values from 0 to 10, and add the even values to a list

In [37]:
num_list = [i for i in range(11) if i%2==0]
print(num_list)

[0, 2, 4, 6, 8, 10]


**Square the number if its even, else, cube it**

In [38]:
[i**2 if i%2==0 else i**3 for i in [2,3,4,5]]

[4, 27, 16, 125]

<a id="nested_condition"> </a>
### 2.4 nested condition
Nested conditions comprise condition statements contained within the definition of other condition statements

**Find all the numbers from a given list divisible by 2 and 3**

In [39]:
numbers = [1, 2, 3, 4, 5, 6, 18, 20]
number_list = [number for number in numbers if number % 2 == 0 if number % 3 == 0]
print(number_list)

[6, 18]


Here, list comprehension checks the number is divisible by 2 and 3. If number satisfies both conditions, number is appended to number_list.

**Return whether the number is even or odd, for each integer from 0 to 20**

In [40]:
["Even" if i%2==0 else "Odd" for i in range(21)]

['Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even']

**Find common numbers from two list using list comprehension**

In [24]:
list_1 = [102, 232, 233, 344, 456, 564, 234]
list_2 = [102, 233, 324, 564, 567, 523, 456]

common_num = [a for a in list_1 for b in list_2 if a == b]

print(common_num)

[102, 233, 456, 564]


<a id="builtin"> </a>
## 3. Built-in Functions

<table align="left">
    <tr>
        <td width="15%">
            <img src="note.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>The Python interpreter has functions instantly available for use, called <i>built-in functions</i>. For example, the print() function prints the given object. <br><br> In Python 3.6 there are 69 built-in function.<br>
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

Some exapmles are: 

<table border="2">
<tr>
<th> Function </th>
<th> Description </th>
</tr>
<tr>
<th> abs()</th>
<th> It gives the absolute of a numeric value </th>
<tr>
<tr>
<th> sum()</th>
<th> It gives the sum of a numeric values values </th>
<tr>
<tr>
<th> len()</th>
<th> It gives the sum of a numeric values values </th>
<tr>   
<tr>
<th> apply()</th>
<th> applied both to series and dataframes where function can be applied both series and individual elements based on the type of function provide </th>
<tr> 
<tr>
<th> map()</th>
<th> only works on a pandas series where type of operation to be applied depends on argument passed as a function, dictionary or a list </th>
<tr>         
</table>    


As we progress with the course, we shall come across a few more built-in functions.

Get the absolute value of a number.

In [1]:
# get the absolute value of -99.34
abs(-99.34)

99.34

A list consists of elements 44, 45, -56, 44, 32,-54, 44, 56, 44, 87, 44, 56. <br>
We can obtain the number of elements in list and sum of all these elements.

In [2]:
x = [44, 45, -56, 44, 32,-54, 44, 56, 44, 87, 44, 56]

In [3]:
# len() is used to check length
len(x)

12

In [4]:
# sum() is used to obtain sum of all the elements in list
sum(x)

386

In [5]:
# create a string
string = "Data Science"

Get the length of the string

In [6]:
# use len() to obtain the length of the string
len(string)

12

The output is the number of characters in the string including the blank space.

Now let us find the number of words in the string.

To do so, we will first break the string at the blank space using the function split() and then with the help of the function len() we obtain the number of words in the string.

In [7]:
len(string.split())

2

Now we create a string 'apple, kiwi, pineapple,mango,grapes, banana'. Is it possible to find the number of fruits in the string.

Yes, it is. We shall use the split function again and then function 'len'. 

In [25]:
# create a string
fruits = 'apple, kiwi, pineapple,mango,grapes, banana'

print(fruits.split())
print("The number of fruits in the list are: " , len(fruits.split()))

['apple,', 'kiwi,', 'pineapple,mango,grapes,', 'banana']
The number of fruits in the list are:  4


It is however evident that there are more than 4 fruits in the list

**Note:** The function split(), breaks the string at a blank space (white space) if no argument is passed. So now we shall specify for it to break at a "," since the string is not spaced properly. 

In [26]:
print(fruits.split(","))
print("The number of fruits in the list are: " , len(fruits.split(",")))

['apple', ' kiwi', ' pineapple', 'mango', 'grapes', ' banana']
The number of fruits in the list are:  6


This gives the correct output.

**Note:** We can specify any character we wish to split at. For instance, a letter, a number, "," , "'" , "." and so on. If nothing is specified, the function will split at a white space by default.

<a id="user"> </a>
## 4. User Defined Functions

There are situations where in built functions are not availabe in a library. In such cases, we do the computations by using formulas/concepts

<table align="left">
    <tr>
        <td width="15%">
            <img src="note.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> At times there are no built-in functions available, thus there arises a need for the user to define a function as desired. The user-defined functions are named blocks of code that are designed to do a specific job(s). When you want to perform a particular task that you’ve defined in a function, you call the name of the function responsible for it. This aids the scenarios where a chunk of code is required repeatedly; just call the function dedicated to handling that task, and the call tells Python to run the code inside the function. You’ll find that using functions makes your programs easier to write, read, test, and fix.
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

### Argument and Parameter


<table align="left">
    <tr>
        <td width="8%">
            <img src="caution.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> Parameters and arguments are subtly different. Lets us first define them.<br><br>
                         1. Parameter: A value which is passed when defining a function
                        <br><br>
                         2. Argument: A value that is passed when invoking a function
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

Consider the following user defined function:

Now, if we invoke this function:

Here the 'sample_parameters' is called argument.

<a id="without_arg"> </a>
## 4.1 User defined functions without arguments

Once a function is defined, it can be called anytime, without running the function again. Generally, a function returns a value or prints a value.  

Let us see a few examples of user defined functions.

**1. We shall write a user defined function without argument to find the twice of a number.**

In [29]:
def twice():
    a = int(input('enter a number: '))
    print(a*2)

In [30]:
# In this function, no argument is passed
# call the function
twice()

44


**2. We shall write a user defined function hi() which take a string input. The output will greet the user displaying his/her name on the screen.**

In [35]:
def hi():
    name = str(input('Enter your name: '))
    print("Hey " + name +", how are you?")

In [36]:
# In this function, no argument is passed
# call the function
hi()

Hey Anurag, how are you?


In [33]:
#create a function
def greetings():
    print("hello")
    print("How r u ?")

In [34]:
# In this function, no argument is passed
# call the function
greetings()

hello
How r u ?


**3. To find area of a circle**

In [27]:
def area_circle():
    r = int(input('Enter the radius of circle: '))
    area = 3.14 * r**2 
    print("The area of cicle with radius " + str(r)  +" is " + str(area) + " sq. units")

In [28]:
# call the function   
area_circle()

The area of cicle with radius 3 is 28.26 sq. units


<a id="with_arg"> </a>
## 4.2 User defined functions with arguments

Most of the function will either `return` a value or `print` an output. Take a note of this in the examples considered below.

**1. A user defined function to get the average of all numbers in a sequence.**

In [37]:
def average(x):
    count = 0
    total = 0
    for i in x:
        count += 1
        total += i
    avg = total/count
    return avg

In [40]:
x = (4, 45, -56, 44, 32,-54, 44, 56, 44, 87, 44, 56)

In [41]:
# call the function   
average(x)

28.833333333333332

The average of the number is 32.1667

If the function requires an argument to be passed and we do not provide it, then an error with be shown.

In [42]:
# call the function   
average()

TypeError: average() missing 1 required positional argument: 'x'

**2. A user defined function with argument to greet a person by name.**

In [43]:
def greet(name):
    name = str(name)
    print("Hello, " + name)

In [44]:
# call the function        
greet('Dave')

Hello, Dave


Enter your name in the paranthese, and your notebook will greet you!.

**Also note that this function prints an output.** 

<table align="left">
    <tr>
        <td width="8%">
            <img src="todo.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> TRY IT YOURSELF: <br> Assign a print() function to a variable. Is it possible?
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

**Answer:** The 'print' output can not be stored in a variable. Moreover, any function which returns a value can be stored in a variable. 

**4. Consider the list given below and write a function to find the unique values of the list**

                [44,45,56,44,32,54,44,56,44,87,44,56]
 

In [45]:
def unique(list1): 
    unique_list = [] 
    for x in list1: 
            if x not in unique_list: 
                 unique_list.append(x)  
    return unique_list    

In [46]:
x = [44,45,56,44,32,54,44,56,44,87,44,56]

# call the function  
# store the output in a variable
unique_values = unique(x)

# print the variable name
unique_values

[44, 45, 56, 32, 54, 87]

These are the unique values in the list.

**Also note that this function returns the output to the calling program.**

**5. A function to reverse a string**

In [51]:
def reverse(string):
    # set the empty string
    rev=''
    for i in string:
        rev = i +rev
    print(rev)

In [52]:
# call the function
string = str(input('Enter string:'))
reverse(string)

erolagnaB


**6. A function to display multiplication table of an integer**

In [29]:
def table(n):
    multi = 1
    for i in range(1,11):
        multi = n*i
        print(str(i) + " times " + str(n) + " is " + str(multi))

In [31]:
# call the function        
n=int(input('Enter number:'))
table(n)

Enter number:23
1 times 23 is 23
2 times 23 is 46
3 times 23 is 69
4 times 23 is 92
5 times 23 is 115
6 times 23 is 138
7 times 23 is 161
8 times 23 is 184
9 times 23 is 207
10 times 23 is 230


**7. To find product of numbers in a list.**

In [32]:
def product(x):
    prod = 1
    for i in x:
        prod *= i
    return prod

In [33]:
#call the function
product([1,2,3,4,5])

120

**8. A function which takes the order of a matrix and mutliplies it with a scalar. The function take two parameters- first the order of identity matrix and second the scaler to be multiplied.**

In [34]:
def scalar_multiplication(n,m):
    for i in range(0,n):
        for j in range(0,n):
            if(i==j):
                print(m,sep=" ",end=" ")
            else:
                print("0",sep=" ",end=" ")
        print()

In [35]:
#call the function
scalar_multiplication(4,5)

5 0 0 0 
0 5 0 0 
0 0 5 0 
0 0 0 5 


**9. Usage of `*args`.**

Usage : multiple arguments can be added without specifying the number of parameter defined in a function.  It is used to pass a non-keyworded, variable-length argument list.

 `*args` in function definitions in python is used to pass a variable number of arguments to a function. It reads the value one by one and performs the required task. <br>
 The symbol `*` is used to take in a variable number of arguments; by convention, it is often used with the word 'args', however we may use another word. 
 

In [55]:
# * indicates that number of arguments can be taken
def Multiple(*args):  
    for parameter in args:  
        print (parameter) 

In [57]:
#call the function
Multiple("Lets","learn","Python","for","Data","Science","and","ML") 

Lets
learn
Python
for
Data
Science
and
ML


**10. Another example on `*args`.**

In [38]:
def to_print( argument1, *argument):
   print(argument1)
   for var in argument:
        print(var)

In [39]:
#call the function        
to_print( 10,20,30 )   

10
20
30


**11. Usage of  ` **kwargs`.**

Usage: `**kwargs` is used to pass a keyworded, variable-length argument list to the function. The name 'kwargs' is used with the double star because it allows us to pass through keyword arguments (and any number of them).

In [40]:
def function(**kwargs):
    for a,b in kwargs.items():
        print(a,"==",b)

In [41]:
# call the function
a=function(A = "First letter", B = "Second letter", C = "Thirs letter")

A == First letter
B == Second letter
C == Thirs letter


**12. Another example of ` **kwargs `.**

`**kwargs` is a dict of the keyword or arguments passed to the function

In [42]:
def print_keyword_args(**kwargs):
      for key, value in kwargs.items():
            print( "%s = %s" % (key, value))

In [43]:
# call the function
print_keyword_args(first_name="John", last_name="Doe")

first_name = John
last_name = Doe


<a id="Recursive_Functions"> </a>
## 4.3 Recursive Functions

A function that calls itself is a recursive function. Recursive function is used when a certain problem is defined in terms of itself.

In [44]:
# calculate the sum of all numbers present in the list
def sum_list(nums):
    if len(nums) == 0:
        return 0

    last_num = nums.pop()
    return last_num + sum_list(nums)

num = [1,2,3,4]
sum_list(num)

10

In [1]:
# Recursive Factorial
def factorial(n):
    if n == 1 or n == 0:
        return 1
    else:
        return n * factorial(n-1)

factorial(5)   

120

In [45]:
# implementation of the fibonacci series using recursion
def fib(n):
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fib(n - 1) + fib(n - 2)
 
 
for i in range(15):
    print(fib(i), end=" ")

0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 

<a id = 'scope'></a>
## 5. Scope of Variables:

<table align="left">
    <tr>
        <td width="15%">
            <img src="note.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> The scope of variables is broadly classified as: <br><br>
                        1. Global Variable: Global variables are the variables declared outside of the function or in global scope. This means, global variable can be accessed inside or outside of the function.
                        <br> <br> <br>
                        2. Local Variable : Local variables are the varibales declared inside the function's body or in the local scope
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

**Following are examples of Global variables:**

The variable 'string' is a global variable.

In [46]:
# This function uses global variable 'string'
def function():  
    print (string)  
  

string = "Python for Data Science"
function()

Python for Data Science


Another example where x is a global variable

In [47]:
def f():
    y = x
    print(y)

In [48]:
x = 8
f()

8


Consider another scenario where x is declared in the function after y is declared.

In [49]:
def f():
    y = x
    print(y)
    x = 22

In [50]:
f()

UnboundLocalError: local variable 'x' referenced before assignment

The output will be an error since x is declared after y and y is x.

### 2. Local Variable

Example of local variables.

In [51]:
def sum(x,y):
    sum = x + y
    return sum

print(sum(4,7)   )

11


Here, x and y are local variables. Their value is not callable outside the definition.

Let us print x and check it value.

In [52]:
x

8

**Here x takes its global value and not the one defined in the function.**

Another example of local variable:

In [53]:
# set a local variable
def local():
    y = "This is a Local variable"
    print(y)

local()

This is a Local variable


Now if we print just y

In [54]:
# local variable y cannot be accessed outside the function
print(y)

NameError: name 'y' is not defined

We are bound to get an error since it. The variable 'y' is a local variable.

<a id = 'lambda' > </a>
## 6. Lambda Functions

<table align="left">
    <tr>
        <td width="15%">
            <img src="note.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> A lambda operator or lambda function is used for creating small, one-time, anonymous function objects in Python.<br><br>The general expression for a lambda function is <br>
                                      lambda argument(s): expression <br><br>
                        A lambda function can take any number of arguments, but they contain only a single expression. An expression is a piece of code executed by the lambda function, which may or may not return any value. <br><br>
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

Let us look at some examples using on lambda functions.

**1. Obtain the minimum among two numbers**

In [3]:
min_value = (lambda x, y: x if x < y else y)
min_value(36, 78)

36

**2.  Check whether two strings are equal**

In [56]:
result = (lambda x, y: 'True' if x == y else 'False')
result("Data", "DaTa")

'False'

**3. Demonstrating wrong usuage of variable name**

In [57]:
# multiplication - wrong usage of variable
x = lambda a, c : a + b
print(x(5, 6))

NameError: name 'b' is not defined

**If variable a and c are used, then the operations should be on a and c.**

<a id='overview'> </a>
## 7. Overview of map, reduce and filter functions

<table align="left">
    <tr>
        <td width="15%">
            <img src="note.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> The function map(f,l) applies a function 'f' to each element of list 'l' <br> 
                        The output of map is not a list, to convert it to a list use list(map(f,l)) <br><br>
                        <br>
                      The function filter(p,l) where property p is checked for every element in list l. The output of fliter(p, l) is a sublist of list 'l' that satistifies the property 'p'. <br><br><br>
                        The reduce(fun,seq) function is used to apply a particular function passed in its argument to all of the list elements mentioned in the sequence passed along and returns only one value.This function is defined in “functools” module. <br><br><br>
                        The  accumulate() in “itertools” module. The syntax for it is  accumulate(l,f). It works similar to the reduce(), however there is a little difference in the output.
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

<a id='map'></a>
### 7.1 Function map()

**1. A test was conducted in a class, the list of all the scores is given. Tha scores are out of 45. Obtain the percentage of each score using lambda function.**
             
             marks = [ 20, 23, 34, 45, 33, 10, 43, 23, 27, 34, 37, 44, 42, 43 ]


In [58]:
marks = [ 20, 23, 34, 45, 33, 10, 43, 23, 27, 34, 37, 44, 42, 43 ]

print(list(map(lambda x, y=100 :x*y/45 , marks)))

[44.44444444444444, 51.111111111111114, 75.55555555555556, 100.0, 73.33333333333333, 22.22222222222222, 95.55555555555556, 51.111111111111114, 60.0, 75.55555555555556, 82.22222222222223, 97.77777777777777, 93.33333333333333, 95.55555555555556]


The output is the percentage scores.

**2. Create two lists and add corresponding elements of each list**

In [59]:
# iteration using list
list_a = [30, 35, 40]
list_b = [12, 5, 70]
  
print(list(map(lambda x, y: x + y, list_a, list_b)))

[42, 40, 110]


<a id='filter'></a>
### 7.2 Function filter()

**1. To obtain the  sum of cubes of odd number from 1 to 50**

In [1]:
# sum of cubes of odd numers from 1 to 50

# user defined function to obtain the cube
def cube(x):
    return(x*x)

# user defined function to obtain odds numbers from a list
def isodd(x):
    return(x%2 == 1)

# create a list of all cubes
list_cubes = list(map(cube,filter(isodd,range(1,50))))

# sum of all elements in the in the list_cube
sum(list_cubes)

20825

The same code above can be done using list comprehension  in the follwing way:

In [2]:
# map - the function defined above is called 
# generator - the range for numbers is fixed
# filter - cube is calculated for the numbers which satisfy the 'if' condition
# built-in function sum() is used to obtain the sum

sum ( [cube(x) for x in range(50) if isodd(x)] ) # Map - Generator - Filter


20825

**2. To obtain the odd numbers from a list**

In [3]:
numbers = [121, 443, 477, 125, 34, 655, 458, 141, 363, 125, 456 , 578, 123]

# function that filters the numbers
def oddnumber(in_num):
    if(in_num % 2) == 0:
        return True
    else:
        return False

# Use filter() to remove odd numbers
out_filter_number = filter(oddnumber, numbers)

print("Type of filter object: ", type(out_filter_number))
print("Even number is as follows: ", list(out_filter_number))

Type of filter object:  <class 'filter'>
Even number is as follows:  [34, 458, 456, 578]


<a id='reduce'></a>
### 7.3 Function reduce()

**1. List of  sales in a shop are noted. Obtain the total sales**

In [4]:
# import reduce function from functools
from functools import reduce

sale = [14565,65460,74450,54546,23245,14543,84355,54566]
total_sale = reduce(lambda a, b: a + b, sale)
print(total_sale)


385730


**Thus we may say the total sale is 385730 INR.**

**2. Factorial of a 5** 

In [5]:
# import reduce function from functools
from functools import reduce

factorial = reduce(lambda p,q : p*q , range(1,7))

print("The factorial of 6 is ",factorial)

The factorial of 6 is  720


<a id='accumulate'></a>
### 7.4 Function accumulate()

**1.  Sum of first ten whole numbers**

In [6]:
# importing itertools for accumulate() 
import itertools 

# priting summation using accumulate() 
print ("The summation of list using accumulate is :",end="") 
print (list(itertools.accumulate(range(10),lambda x,y : x+y))) 

The summation of list using accumulate is :[0, 1, 3, 6, 10, 15, 21, 28, 36, 45]


The output of reduce() will be:

In [7]:
# import reduce function from functools
from functools import reduce

sum = reduce(lambda a,b : a+b , range(10))
print(sum)

45


The function accumulate() displays the iterations whereas the function reduce() displays the final output.